swarm/storage/localstore/mode.go

```diff
@@ -0,0 +1,182 @@
+func (db *DB) access(mode Mode, item shed.IndexItem) (out shed.IndexItem, err error) {
+	out, err = db.retrievalIndex.Get(item)
+	if err != nil {
+		return out, err
+	}
+	switch mode {
+	case ModeRequest:
+		// update the access counter
+		// Q: can we do this asynchronously
+		return out, db.update(context.TODO(), mode, item)
+	default:
+		// all other modes are not updating the index
+	}
+	return out, nil
+}
+
+// update is called by an Accessor with a specific Mode.
+// This function calles updateBatch to perform operations
+// on indexes and fields within a single batch.
+func (db *DB) update(ctx context.Context, mode Mode, item shed.IndexItem) error {
+	db.mu.RLock()
+	b := db.batch
+	db.mu.RUnlock()
+
+	// check if the database is not closed
+	select {
+	case <-db.close:
+		return ErrDBClosed
+	default:
+	}
+
+	// call the update with the provided mode
+	err := db.updateBatch(b, mode, item)
+	if err != nil {
+		return err
+	}
+	// trigger the writeBatches loop
+	select {
+	case db.writeTrigger <- struct{}{}:
+	default:
+	}
+	// wait for batch to be written and return batch error
+	// this is in order for Put calls to be synchronous
+	select {
+	case <-b.Done:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return b.Err
+}
+
+// batch wraps leveldb.Batch extending it with a done channel.
+type batch struct {
+	*leveldb.Batch
+	Done chan struct{} // to signal when batch is written
+	Err  error         // error resulting from write
+}
+
+// newBatch constructs a new batch.
+func newBatch() *batch {
+	return &batch{
+		Batch: new(leveldb.Batch),
+		Done:  make(chan struct{}),
+	}
+}
+
+// updateBatch performs different operations on fields and indexes
+// depending on the provided Mode.
+func (db *DB) updateBatch(b *batch, mode Mode, item shed.IndexItem) (err error) {
+	switch mode {
+	case ModeSyncing:
+		// put to indexes: retrieve, pull
+		item.StoreTimestamp = now()
+		item.AccessTimestamp = now()
+		db.retrievalIndex.PutInBatch(b.Batch, item)
+		db.pullIndex.PutInBatch(b.Batch, item)
+		db.sizeCounter.IncInBatch(b.Batch)
+
+	case ModeUpload:
+		// put to indexes: retrieve, push, pull
+		item.StoreTimestamp = now()
+		item.AccessTimestamp = now()
+		db.retrievalIndex.PutInBatch(b.Batch, item)
+		db.pullIndex.PutInBatch(b.Batch, item)
+		db.pushIndex.PutInBatch(b.Batch, item)
+
+	case ModeRequest:
+		// put to indexes: retrieve, gc
+		item.StoreTimestamp = now()
+		item.AccessTimestamp = now()
+		db.retrievalIndex.PutInBatch(b.Batch, item)
+		db.gcIndex.PutInBatch(b.Batch, item)
+
+	case ModeSynced:
+		// delete from push, insert to gc
+		item.StoreTimestamp = now()
+		db.retrievalIndex.PutInBatch(b.Batch, item)
+		db.pushIndex.DeleteInBatch(b.Batch, item)
+		db.gcIndex.PutInBatch(b.Batch, item)
+
+	case ModeAccess:
+		// update accessTimeStamp in retrieve, gc
+		db.gcIndex.DeleteInBatch(b.Batch, item)
+		item.AccessTimestamp = now()
+		db.retrievalIndex.PutInBatch(b.Batch, item)
+		db.gcIndex.PutInBatch(b.Batch, item)
+
+	case ModeRemoval:
```

Comment:
`sizeCounter` should be decremented here right?

Reply:
Yes, you are right.